src/server/pps/server/api_server.go

```diff
@@ -3559,3 +3560,80 @@ func newMessageFilterFunc(jqFilter string, projects []*pfs.Project) (func(contex
 		return true, nil
 	}, nil
 }
+
+// emptyPipelineSpecJSON is the result of marshalling a zero pps.PipelineSpec
+// value.  It is used in case any of the PipelineSpec fields are marshaled even
+// if empty.
+var emptyPipelineSpecJSON string
+
+func init() {
+	var spec pps.PipelineSpec
+	b, err := json.Marshal(spec)
+	if err != nil {
+		panic(fmt.Sprint("could not marshal empty pipeline spec: ", err))
+	}
+	emptyPipelineSpecJSON = string(b)
+}
+
+func (a *apiServer) GetClusterDefaults(ctx context.Context, req *pps.GetClusterDefaultsRequest) (*pps.GetClusterDefaultsResponse, error) {
+	var clusterDefaults pps.ClusterDefaults
+	if err := a.clusterDefaults.ReadOnly(ctx).Get("", &clusterDefaults); err != nil {
+		if !errors.As(err, &col.ErrNotFound{}) {
+			return nil, errors.Wrap(err, "could not read cluster defaults")
+		}
+		clusterDefaults.DetailsJson = "{}"
+	}
+
+	var err error
+	if clusterDefaults.EffectiveDetailsJson, err = jsonMergePatch(emptyPipelineSpecJSON, clusterDefaults.DetailsJson); err != nil {
+		return nil, errors.Wrap(err, "could not merge empty spec with cluster default details")
+	}
+
+	return &pps.GetClusterDefaultsResponse{ClusterDefaults: &clusterDefaults}, nil
+}
+
+// jsonMergePatch merges a JSON patch in string form with a JSON target, also in
+// string form.
+func jsonMergePatch(target, patch string) (string, error) {
+	var targetObject, patchObject any
+	if err := json.Unmarshal([]byte(target), &targetObject); err != nil {
+		return "", errors.Wrap(err, "could not unmarshal target JSON")
+	}
+	if err := json.Unmarshal([]byte(patch), &patchObject); err != nil {
+		return "", errors.Wrap(err, "could not unmarshal patch JSON")
+	}
+	result, err := json.Marshal(mergePatch(targetObject, patchObject))
+	if err != nil {
+		return "", errors.Wrap(err, "could not marshal merge patch result")
+	}
+	return string(result), nil
+}
+
+// mergePatch implements the RFC 7396 algorithm.  To quote the RFC “If the patch
+// is anything other than an object, the result will always be to replace the
```

Comment:
Why do we allow anything other than an object?  Why not unmarshal into `pps.PipelineSpec` instead of `any`?

Reply:
> Why do we allow anything other than an object?

It’s the recursive algorithm defined by the RFC, so it doesn’t only get called on objects, but also on their values (at `targetMap[name] = mergePatch(targetMap[name], value)`), which may be of any type.

>Why not unmarshal into `pps.PipelineSpec` instead of `any`?

It may be called on a JSON object which represents a pipeline spec, but it may also be called on a JSON object which represents a resource request, or a toleration … or it may be called on a string/bool/number/array/nil.

I could implement this using reflection instead, but I don’t know if that’s a huge advantage.

Reply:
E.g. with a JSON merge, `mergePatch({"resource_limits": {"cpu": "2m"}, "resource_requests": {"cpu": "1m"}, {"resource_request": nil}) → {"resource_limits": {"cpu": "2m"}}`, but with a Go-typed, reflection-based approach it would yield `{}` because `"resource_limits"}` was omitted from the patch spec.

Reply:
I think the embedded messages that need a custom type to detect presence then:

```go
type ResourceRequests struct {
   Present bool `json:"_"`
   CPU int `json:"cpu"`
}
```
func (req *ResourceRequests) UnmarshalJSON(buf []byte) error {
   if len(buf) == 0 {
      return nil
   }
   req.Present = true
   var result struct { CPU int `json:"cpu"` }
   if err := json.Unmarshal(buf, &result); err != nil {
       return errors.Wrap(err, "unmarshal resource requests")
   }
   req.CPU = result.CPU
   return nil
}
func (req *ResourceRequests) MarshalJSON() ([]byte, error) {
    if !req.Present { return nil, nil }
    json.Marshal(...)
    ...
}
```
This is a lot of boilerplate to write, but allows invalid requests to be aborted early with a descriptive message.

Reply:
It turns out that your initial hunch over at https://github.com/pachyderm/pachyderm/pull/9010#discussion_r1247209379 was correct: this need to be `protojson.Marshal`able, which means that the struct must be a protobuf message, which means that it’s not possible to add `Present` fields to it.

Reply:
We talked about this in person and the any type seems fine.